File: user-service/src/main/java/com/athena/userservice/service/AuthService.java
package com.athena.userservice.service;

import com.athena.userservice.config.JwtUtil;
import com.athena.userservice.dto.AuthRequest;
import com.athena.userservice.dto.AuthResponse;
import com.athena.userservice.model.User;
import com.athena.userservice.repository.RoleRepository;
import com.athena.userservice.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class AuthService {

    private final UserRepository userRepository;
    private final RoleRepository roleRepository;
    private final PasswordEncoder passwordEncoder;
    private final JwtUtil jwtUtil;
    private final PasswordPolicyService passwordPolicyService;

    public AuthResponse register(AuthRequest request) {
        if (userRepository.findByUsername(request.getUsername()).isPresent()) {
            throw new RuntimeException("Username already exists");
        }
        passwordPolicyService.validatePassword(request.getPassword());

        var defaultRole = roleRepository.findByName("USER")
                .orElseThrow(() -> new RuntimeException("Default role USER not found"));

        User user = User.builder()
                .username(request.getUsername())
                .password(passwordEncoder.encode(request.getPassword()))
                .status("ACTIVE")
                .build();
        user.getRoles().add(defaultRole);
        userRepository.save(user);

        List<String> roles = user.getRoles().stream().map(r -> r.getName()).collect(Collectors.toList());
        String token = jwtUtil.generateToken(user.getUsername(), roles, null);

        return AuthResponse.builder()
                .token(token)
                .username(user.getUsername())
                .roles(roles)
                .build();
    }

    public AuthResponse authenticate(AuthRequest request) {
        User user = userRepository.findByUsername(request.getUsername())
                .orElseThrow(() -> new RuntimeException("Invalid credentials"));

        if (!passwordEncoder.matches(request.getPassword(), user.getPassword())) {
            throw new RuntimeException("Invalid credentials");
        }

        List<String> roles = user.getRoles().stream().map(r -> r.getName()).collect(Collectors.toList());
        List<String> groups = user.getGroups().stream().map(g -> g.getName()).collect(Collectors.toList());
        String token = jwtUtil.generateToken(user.getUsername(), roles, null);

        return AuthResponse.builder()
                .token(token)
                .username(user.getUsername())
                .firstName(user.getFirstName())
                .lastName(user.getLastName())
                .email(user.getEmail())
                .roles(roles)
                .groups(groups)
                .build();
    }
}

File: user-service/internal/service/auth_service.go
package service

import (
	"fmt"

	"github.com/athena/pkg/jwt"
	"github.com/athena/user-service/internal/dto"
	"github.com/athena/user-service/internal/repository"
	"golang.org/x/crypto/bcrypt"
)

type AuthService struct {
	userRepo      *repository.UserRepository
	roleRepo      *repository.RoleRepository
	jwtUtil       *jwt.JWTUtil
	policyService *PasswordPolicyService
}

func NewAuthService(
	userRepo *repository.UserRepository,
	roleRepo *repository.RoleRepository,
	jwtUtil *jwt.JWTUtil,
	policyService *PasswordPolicyService,
) *AuthService {
	return &AuthService{
		userRepo:      userRepo,
		roleRepo:      roleRepo,
		jwtUtil:       jwtUtil,
		policyService: policyService,
	}
}

// Authenticate validates credentials against the users table and returns a JWT.
func (s *AuthService) Authenticate(req dto.AuthRequest) (*dto.AuthResponse, error) {
	user, err := s.userRepo.FindByUsername(req.Username)
	if err != nil {
		return nil, fmt.Errorf("Invalid credentials")
	}

	if err := bcrypt.CompareHashAndPassword([]byte(user.Password), []byte(req.Password)); err != nil {
		return nil, fmt.Errorf("Invalid credentials")
	}

	roles := make([]string, 0, len(user.Roles))
	for _, r := range user.Roles {
		roles = append(roles, r.Name)
	}
	groups := make([]string, 0, len(user.Groups))
	for _, g := range user.Groups {
		groups = append(groups, g.Name)
	}

	token, err := s.jwtUtil.GenerateToken(user.Username, roles, nil)
	if err != nil {
		return nil, fmt.Errorf("failed to generate token: %w", err)
	}

	return &dto.AuthResponse{
		Token:     token,
		Username:  user.Username,
		FirstName: user.FirstName,
		LastName:  user.LastName,
		Email:     user.Email,
		Roles:     roles,
		Groups:    groups,
	}, nil
}
